File: helpers/twitter.py
import re
import tweepy
from boto3 import client
from typing import Dict

from helpers.images import fit_image_to_constraint, get_image, to_bytes
from helpers.wikimedia import get_file_details

ssm = client('ssm')
consumer_key = ssm.get_parameter(Name='/wikimedia_bot/twitter/api_key')['Parameter']['Value']
consumer_secret = ssm.get_parameter(Name='/wikimedia_bot/twitter/api_key_secret')['Parameter']['Value']
access_token = ssm.get_parameter(Name='/wikimedia_bot/twitter/access_token')['Parameter']['Value']
access_token_secret = ssm.get_parameter(Name='/wikimedia_bot/twitter/access_token_secret')['Parameter']['Value']

auth = tweepy.OAuth1UserHandler(consumer_key, consumer_secret, access_token, access_token_secret)
api = tweepy.API(auth)
client = tweepy.Client(consumer_key=consumer_key, consumer_secret=consumer_secret,
                       access_token=access_token, access_token_secret=access_token_secret)

def post(file: Dict[str, any]) -> None:
  try:
    details = get_file_details(file['title'])
    alt_text = details['extmetadata']['ImageDescription']['value']
    alt_text = re.sub('<[^<]+?>', '', alt_text)
    image = get_image(details['url'])
    resized_image  = fit_image_to_constraint(image, 2048)
    media_upload = api.simple_upload(f'image.{image.format}', file=to_bytes(resized_image))
    api.create_media_metadata(media_upload.media_id, alt_text=alt_text)
    client.create_tweet(media_ids=[media_upload.media_id])
  except Exception as e:
    print('Error posting to Twitter')
    print(e)

File: src/helpers/twitter.ts
import { SSMClient, GetParameterCommand } from '@aws-sdk/client-ssm';
import { TwitterApi } from 'twitter-api-v2';
import { WikimediaFile } from '../types';
import { getFileDetails, getImage, fitImageToConstraint, toBuffer } from './';

let twitterClient: TwitterApi | undefined;

const getTwitterClient = async (): Promise<TwitterApi> => {
  if (!twitterClient) {
    const ssmClient = new SSMClient({});
    const [apiKey, apiSecret, accessToken, accessSecret] = await Promise.all([
      ssmClient.send(
        new GetParameterCommand({
          Name: '/wikimedia_bot/twitter/api_key',
        })
      ),
      ssmClient.send(
        new GetParameterCommand({
          Name: '/wikimedia_bot/twitter/api_key_secret',
        })
      ),
      ssmClient.send(
        new GetParameterCommand({
          Name: '/wikimedia_bot/twitter/access_token',
        })
      ),
      ssmClient.send(
        new GetParameterCommand({
          Name: '/wikimedia_bot/twitter/access_token_secret',
        })
      ),
    ]);

    if (
      !apiKey.Parameter?.Value ||
      !apiSecret.Parameter?.Value ||
      !accessToken.Parameter?.Value ||
      !accessSecret.Parameter?.Value
    ) {
      throw new Error('twitter credentials not found in ssm');
    }

    twitterClient = new TwitterApi({
      appKey: apiKey.Parameter.Value,
      appSecret: apiSecret.Parameter.Value,
      accessToken: accessToken.Parameter.Value,
      accessSecret: accessSecret.Parameter.Value,
    });
  }
  return twitterClient;
};

export const post = async (file: WikimediaFile): Promise<void> => {
  try {
    const twitter = await getTwitterClient();
    const details = await getFileDetails(file.title);
    const image = await getImage(details.url);
    const resizedImage = await fitImageToConstraint(image, 2048);
    const imageBuffer = await toBuffer(resizedImage);

    const mediaId = await twitter.v1.uploadMedia(imageBuffer, {
      mimeType: 'image/jpeg',
    });

    await twitter.v1.createMediaMetadata(mediaId, {
      alt_text: { text: details.descriptionurl },
    });

    await twitter.v2.tweet('', {
      media: { media_ids: [mediaId] },
    });
  } catch (e) {
    console.error('error posting to twitter:', e);
    throw e;
  }
};
